File: api/src/jira-integrations/persistence/dailyHoursRepository.ts
import { DataSource } from 'typeorm';
import { dataSource as defaultDataSource } from 'database/createConnection';
import DailyHoursEntity from 'entities/DailyHours';

export type HoursSource = 'tempo' | 'manual';

export interface DailyHoursRow {
  id: number;
  workDate: string;
  hoursLogged: number;
  source: HoursSource;
  createdAt: Date;
  updatedAt: Date;
}

export interface IDailyHoursRepository {
  getByDate(workDate: string): Promise<DailyHoursRow | null>;
  getWeekHours(startDate: string, endDate: string): Promise<DailyHoursRow[]>;
  upsert(workDate: string, hours: number, source: HoursSource): Promise<DailyHoursRow>;
  delete(workDate: string): Promise<boolean>;
}

export class DailyHoursRepository implements IDailyHoursRepository {
  private dataSource: DataSource;

  constructor(dataSource: DataSource = defaultDataSource) {
    this.dataSource = dataSource;
  }

  private get repo() {
    return this.dataSource.getRepository(DailyHoursEntity);
  }

  /**
   * Get hours for a specific date
   */
  async getByDate(workDate: string): Promise<DailyHoursRow | null> {
    const entity = await this.repo.findOne({ where: { workDate } });
    return entity ? this.toRow(entity) : null;
  }

  /**
   * Get hours for a date range (e.g., Monday to Sunday)
   */
  async getWeekHours(startDate: string, endDate: string): Promise<DailyHoursRow[]> {
    const entities = await this.repo
      .createQueryBuilder('d')
      .where('d.work_date >= :startDate', { startDate })
      .andWhere('d.work_date <= :endDate', { endDate })
      .orderBy('d.work_date', 'ASC')
      .getMany();
    return entities.map((e) => this.toRow(e));
  }

  /**
   * Insert or update hours for a date
   */
  async upsert(workDate: string, hours: number, source: HoursSource): Promise<DailyHoursRow> {
    const repo = this.repo;
    const existing = await repo.findOne({ where: { workDate } });

    if (existing) {
      existing.hoursLogged = hours;
      existing.source = source;
      const saved = await repo.save(existing);
      return this.toRow(saved);
    } else {
      const entity = repo.create({ workDate, hoursLogged: hours, source });
      const saved = await repo.save(entity);
      return this.toRow(saved);
    }
  }

  /**
   * Delete hours for a date
   * @returns true if a record was deleted, false otherwise
   */
  async delete(workDate: string): Promise<boolean> {
    const result = await this.repo.delete({ workDate });
    return (result.affected ?? 0) > 0;
  }

  private toRow(entity: DailyHoursEntity): DailyHoursRow {
    return {
      id: entity.id,
      workDate: entity.workDate,
      hoursLogged: parseFloat(String(entity.hoursLogged)),
      source: entity.source as HoursSource,
      createdAt: entity.createdAt,
      updatedAt: entity.updatedAt,
    };
  }
}

File: packages/planny-jira/src/planny_jira/repositories/daily_hours_repository.py
"""Repository for DailyHours.

Mirrors ``api/src/jira-integrations/persistence/dailyHoursRepository.ts``.
"""

from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from planny_core.models import DailyHours


class DailyHoursRepository:
    """CRUD operations for daily hours tracking."""

    async def get_by_date(
        self,
        db: AsyncSession,
        work_date: str,
    ) -> DailyHours | None:
        """Find a daily hours record by exact date."""
        stmt = select(DailyHours).where(DailyHours.work_date == work_date)
        result = await db.execute(stmt)
        return result.scalars().first()

    async def upsert_hours(
        self,
        db: AsyncSession,
        work_date: str,
        hours_logged: float,
        source: str = "tempo",
    ) -> DailyHours:
        """Insert or update daily hours for a date."""
        stmt = select(DailyHours).where(DailyHours.work_date == work_date)
        result = await db.execute(stmt)
        existing = result.scalars().first()

        if existing:
            existing.hours_logged = hours_logged
            existing.source = source
            existing.updated_at = datetime.now(timezone.utc)
            record = existing
        else:
            record = DailyHours(
                work_date=work_date,
                hours_logged=hours_logged,
                source=source,
            )
            db.add(record)

        await db.flush()
        return record

    async def get_week_hours(
        self,
        db: AsyncSession,
        start_date: str,
        end_date: str,
    ) -> list[DailyHours]:
        """Return records within a date range (inclusive), ordered by date ASC."""
        stmt = (
            select(DailyHours)
            .where(DailyHours.work_date >= start_date)
            .where(DailyHours.work_date <= end_date)
            .order_by(DailyHours.work_date.asc())
        )
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def delete(
        self,
        db: AsyncSession,
        work_date: str,
    ) -> bool:
        """Delete a daily hours record. Returns True if anything was deleted."""
        stmt = select(DailyHours).where(DailyHours.work_date == work_date)
        result = await db.execute(stmt)
        existing = result.scalars().first()
        if existing is None:
            return False
        await db.delete(existing)
        await db.flush()
        return True
